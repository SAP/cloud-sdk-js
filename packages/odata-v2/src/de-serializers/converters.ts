import moment from 'moment';
import { durationRegexV2 } from '@sap-cloud-sdk/odata-common/internal';
import { Time } from '@sap-cloud-sdk/odata-common';

/**
 * @internal
 */
export function deserializeToTime(value: string): Time {
  const regexResult = durationRegexV2.exec(value);
  if (!regexResult) {
    throw new Error(`Failed to parse the value: ${value} to time.`);
  }
  const { hours, minutes, seconds } = regexResult?.groups || {};

  return {
    hours: hours ? parseInt(hours.replace('H', ''), 10) : 0,
    minutes: minutes ? parseInt(minutes.replace('M', ''), 10) : 0,
    seconds: seconds ? parseInt(seconds.replace('S', ''), 10) : 0
  };
}

/**
 * @internal
 */
export function serializeFromTime(value: Time): string {
  return (
    'PT' +
    leftpad(value.hours, 2) +
    'H' +
    leftpad(value.minutes, 2) +
    'M' +
    leftpad(value.seconds, 2) +
    'S'
  );
}

/**
 * @internal
 * This function can be used for both Edm.DateTime and and Edm.DateTimeOffset.
 */
export function deserializeToMoment(edmDateTime: string): moment.Moment {
  const dateTimeOffsetComponents =
    /^\/Date\((?<signedticks>-{0,1}\d+)((?<offsetsign>[+-])(?<unsignedoffset>\d{4}))?\)\/$/.exec(
      edmDateTime
    )?.groups;
  if (!dateTimeOffsetComponents) {
    throw new Error(`Failed to parse edmDateTime: ${edmDateTime} to moment.`);
  }

  const timestamp = moment(parseInt(dateTimeOffsetComponents.signedticks));

  if (
    dateTimeOffsetComponents.offsetsign &&
    dateTimeOffsetComponents.unsignedoffset
  ) {
    const offsetMultiplier =
      dateTimeOffsetComponents.offsetsign === '+' ? 1 : -1;
    const offsetInMinutes = parseInt(dateTimeOffsetComponents.unsignedoffset);
    return timestamp.utc().utcOffset(offsetMultiplier * offsetInMinutes);
  }

  return timestamp;
}

/**
 * @internal
 * This function can be used for both Edm.DateTime and and Edm.DateTimeOffset.
 */
export function serializeFromMoment(momentInstance: moment.Moment): string {
  const timestamp = momentInstance.unix() * 1000;

  // For some reason isUtc() returns wrong values here, so we use the internal flag directly
  if (momentInstance['_isUTC']) {
    const offset = Math.abs(momentInstance.utcOffset());
    const operator = momentInstance.utcOffset() >= 0 ? '+' : '-';
    return `/Date(${timestamp}${operator}${leftpad(offset, 4)})/`;
  }

  return `/Date(${timestamp})/`;
}

function leftpad(value: any, targetLength: number): string {
  const str = value.toString();
  if (str.length >= targetLength) {
    return str;
  }
  return '0'.repeat(targetLength - str.length) + str;
}
