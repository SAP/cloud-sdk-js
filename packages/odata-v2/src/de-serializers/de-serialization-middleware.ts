import BigNumber from 'bignumber.js';
import {
  DeSerializationMiddleware as DeSerializationMiddlewareBase,
  DeSerializer,
  EdmTypeCommon,
  ExclusiveEdmTypeV2,
  Time
} from '@sap-cloud-sdk/odata-common';
import moment from 'moment';

export interface DeSerializationMiddleware<
  BinaryT = string,
  BooleanT = boolean,
  ByteT = number,
  DecimalT = BigNumber,
  DoubleT = number,
  FloatT = number,
  Int16T = number,
  Int32T = number,
  Int64T = BigNumber,
  GuidT = string,
  SByteT = number,
  SingleT = number,
  StringT = string,
  AnyT = any,
  DateTimeT = moment.Moment,
  DateTimeOffsetT = moment.Moment,
  TimeT = Time
> extends DeSerializationMiddlewareBase<
    BinaryT,
    BooleanT,
    ByteT,
    DecimalT,
    DoubleT,
    FloatT,
    Int16T,
    Int32T,
    Int64T,
    GuidT,
    SByteT,
    SingleT,
    StringT,
    AnyT
  > {
  'Edm.DateTime': DeSerializer<DateTimeT>;
  'Edm.DateTimeOffset': DeSerializer<DateTimeOffsetT>;
  'Edm.Time': DeSerializer<TimeT>;
}

export type DeSerializationMiddlewareV2BASE = {
  [P in EdmTypeCommon | ExclusiveEdmTypeV2]: DeSerializer<any>;
};
