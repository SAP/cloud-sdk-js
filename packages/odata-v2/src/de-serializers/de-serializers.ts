import {
  DeSerializer,
  DeSerializers as DeSerializersCommon
} from '@sap-cloud-sdk/odata-common';

/**
 * Represents a set of functions that determine (de-)serialization per EDM type.
 */
export interface DeSerializers<
  BinaryT = any,
  BooleanT = any,
  ByteT = any,
  DecimalT = any,
  DoubleT = any,
  FloatT = any,
  Int16T = any,
  Int32T = any,
  Int64T = any,
  GuidT = any,
  SByteT = any,
  SingleT = any,
  StringT = any,
  AnyT = any,
  DateTimeT = any,
  DateTimeOffsetT = any,
  TimeT = any
> extends DeSerializersCommon<
    BinaryT,
    BooleanT,
    ByteT,
    DecimalT,
    DoubleT,
    FloatT,
    Int16T,
    Int32T,
    Int64T,
    GuidT,
    SByteT,
    SingleT,
    StringT,
    AnyT
  > {
  /**
   * DeSerializer for `Edm.DateTime` to the generic type `DateTimeT`.
   */
  'Edm.DateTime': DeSerializer<DateTimeT>;
  /**
   * DeSerializer for `Edm.DateTimeOffset` to the generic type `DateTimeOffsetT`.
   */
  'Edm.DateTimeOffset': DeSerializer<DateTimeOffsetT>;
  /**
   * DeSerializer for `Edm.Time` to the generic type `TimeT`.
   */
  'Edm.Time': DeSerializer<TimeT>;
}
