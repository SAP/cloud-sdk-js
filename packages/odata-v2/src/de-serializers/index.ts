export * from './converters';
export * from './custom-de-serializers';
export * from './de-serializers';
export * from './default-de-serializers';
export * from './entity-deserializer';
export * from './payload-value-converter';
