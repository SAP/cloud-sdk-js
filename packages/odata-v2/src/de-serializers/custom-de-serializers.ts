import {
  Time,
  CustomOrDefaultType as CustomOrDefaultTypeCommon
} from '@sap-cloud-sdk/odata-common';
import BigNumber from 'bignumber.js';
import { DeSerializers } from './de-serializers';
import {
  DefaultDeSerializers,
  defaultDeSerializers
} from './default-de-serializers';

/**
 * Get a complete set of (de-)serializers, that consists of the given partial custom (de-)serializers and default (de-)serializers (aka. default (de-)serializers merged with custom (de-)serializers).
 * The custom (de-)serializers are merged with the default (de-)serializers, while custom (de-)serializers take precedence.
 * @param customDeSerializers - Custom (de-)serialization functions.
 * @returns A full set of (de-)serialization functions, that combines custom and default (de-)serializers.
 */
export function mergeDefaultDeSerializersWith<
  BinaryT = string,
  BooleanT = boolean,
  ByteT = number,
  DecimalT = BigNumber,
  DoubleT = number,
  FloatT = number,
  Int16T = number,
  Int32T = number,
  Int64T = BigNumber,
  GuidT = string,
  SByteT = number,
  SingleT = number,
  StringT = string,
  AnyT = any,
  DateTimeT = moment.Moment,
  DateTimeOffsetT = moment.Moment,
  TimeT = Time
>(
  customDeSerializers: Partial<
    DeSerializers<
      BinaryT,
      BooleanT,
      ByteT,
      DecimalT,
      DoubleT,
      FloatT,
      Int16T,
      Int32T,
      Int64T,
      GuidT,
      SByteT,
      SingleT,
      StringT,
      AnyT,
      DateTimeT,
      DateTimeOffsetT,
      TimeT
    >
  >
): CustomDeSerializers<typeof customDeSerializers> {
  return {
    ...(defaultDeSerializers as any),
    ...(customDeSerializers || {})
  };
}

/**
 * Infers the deserialized type for an EDM type, based on custom (de-)serializers.
 * If the custom (de-)serializers specify a type for the given EDM type, this type is inferred.
 * Otherwise the given DefaultType is used.
 * @typeParam CustomDeSerializerT - Type of the custom (de-)serializers.
 * @typeParam EdmT - The EDM type to infer the type for.
 */
export type CustomOrDefaultType<CustomDeSerializerT, EdmT> =
  CustomOrDefaultTypeCommon<CustomDeSerializerT, EdmT, DefaultDeSerializers>;

/**
 * Type of the full set of (de-)serialization functions, that include custom (de-)serializers (aka. default (de-)serializers type merged with custom (de-)serializers type).
 */
export type CustomDeSerializers<T> = DeSerializers<
  CustomOrDefaultType<T, 'Edm.Binary'>,
  CustomOrDefaultType<T, 'Edm.Boolean'>,
  CustomOrDefaultType<T, 'Edm.Byte'>,
  CustomOrDefaultType<T, 'Edm.Decimal'>,
  CustomOrDefaultType<T, 'Edm.Double'>,
  CustomOrDefaultType<T, 'Edm.Float'>,
  CustomOrDefaultType<T, 'Edm.Int16'>,
  CustomOrDefaultType<T, 'Edm.Int32'>,
  CustomOrDefaultType<T, 'Edm.Int64'>,
  CustomOrDefaultType<T, 'Edm.Guid'>,
  CustomOrDefaultType<T, 'Edm.SByte'>,
  CustomOrDefaultType<T, 'Edm.Single'>,
  CustomOrDefaultType<T, 'Edm.String'>,
  CustomOrDefaultType<T, 'Edm.Any'>,
  CustomOrDefaultType<T, 'Edm.DateTime'>,
  CustomOrDefaultType<T, 'Edm.DateTimeOffset'>,
  CustomOrDefaultType<T, 'Edm.Time'>
>;
