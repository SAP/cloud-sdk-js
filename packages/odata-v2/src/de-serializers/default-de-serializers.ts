import {
  defaultDeSerializersRaw as defaultDeSerializersCommon,
  Time,
  wrapDefaultDeSerializers
} from '@sap-cloud-sdk/odata-common/internal';
import BigNumber from 'bignumber.js';
import {
  deserializeToMoment,
  deserializeToTime,
  serializeFromMoment,
  serializeFromTime
} from './converters';
import { DeSerializers } from './de-serializers';

/**
 * Type of the default (de-)serializers.
 */
export type DefaultDeSerializers = DeSerializers<
  string,
  boolean,
  number,
  BigNumber,
  number,
  number,
  number,
  number,
  BigNumber,
  string,
  number,
  number,
  string,
  any,
  moment.Moment,
  moment.Moment,
  Time
>;

/**
 * Default (de-)serializers without `null` and `undefined` handling.
 */
const defaultDeSerializersRaw: DefaultDeSerializers = {
  ...defaultDeSerializersCommon,
  'Edm.DateTime': {
    deserialize: deserializeToMoment,
    serialize: serializeFromMoment,
    serializeToUri: value =>
      `datetime'${value.toISOString().replace(/Z$/, '')}'`
  },
  'Edm.DateTimeOffset': {
    deserialize: deserializeToMoment,
    serialize: serializeFromMoment,
    serializeToUri: value => `datetimeoffset'${value.toISOString()}'`
  },
  'Edm.Time': {
    deserialize: deserializeToTime,
    serialize: serializeFromTime,
    serializeToUri: (value, serialize) => `time'${serialize(value)}'`
  },
  /* DeSerializers with v2 specific URI serializer defaults. */
  'Edm.Decimal': {
    ...defaultDeSerializersCommon['Edm.Decimal'],
    serializeToUri: (value, serialize) => `${serialize(value)}M`
  },
  'Edm.Guid': {
    ...defaultDeSerializersCommon['Edm.Guid'],
    serializeToUri: (value, serialize) => `guid'${serialize(value)}'`
  }
};

/**
 * The default (de-)serializers.
 */
export const defaultDeSerializers = wrapDefaultDeSerializers(
  defaultDeSerializersRaw
);
