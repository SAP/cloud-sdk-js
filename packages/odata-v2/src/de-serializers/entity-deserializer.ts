import {
  entityDeserializer as entityDeserializerBase,
  EntityDeserializer
} from '@sap-cloud-sdk/odata-common';
import { getLinkedCollectionResult } from '../request-builder/response-data-accessor';
import { extractODataEtag } from '../extract-odata-etag';
import { DeSerializers } from './de-serializers';

/**
 * Entity deserializer instance for v2 entities.
 * See {@link @sap-cloud-sdk/odata-common!EntityDeserializer} for the provided methods.
 * @param deSerializers - DeSerializer.
 * @returns EntityDeserializer.
 */
export function entityDeserializer<DeSerializersT extends DeSerializers>(
  deSerializers: DeSerializersT
): EntityDeserializer {
  return entityDeserializerBase(
    deSerializers,
    extractODataEtag,
    getLinkedCollectionResult
  );
}
