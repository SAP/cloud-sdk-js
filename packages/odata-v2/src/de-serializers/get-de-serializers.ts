import { Time } from '@sap-cloud-sdk/odata-common/internal';
import {
  DeSerializationMiddlewareBASE,
  DeSerializer
} from '@sap-cloud-sdk/odata-common/src/de-serializers/de-serialization-middleware';
import BigNumber from 'bignumber.js';
import { DeSerializationMiddleware } from './de-serialization-middleware';
import { defaultDeSerializers } from './default-de-serializers';

export function getDeSerializers<
  BinaryT = string,
  BooleanT = boolean,
  ByteT = number,
  DecimalT = BigNumber,
  DoubleT = number,
  FloatT = number,
  Int16T = number,
  Int32T = number,
  Int64T = BigNumber,
  GuidT = string,
  SByteT = number,
  SingleT = number,
  StringT = string,
  AnyT = any,
  DateTimeT = moment.Moment,
  DateTimeOffsetT = moment.Moment,
  TimeT = Time
>(
  deSerializers: Partial<
    DeSerializationMiddleware<
      BinaryT,
      BooleanT,
      ByteT,
      DecimalT,
      DoubleT,
      FloatT,
      Int16T,
      Int32T,
      Int64T,
      GuidT,
      SByteT,
      SingleT,
      StringT,
      AnyT,
      DateTimeT,
      DateTimeOffsetT,
      TimeT
    >
  >
): CustomDeSerializer<typeof deSerializers> {
  return {
    ...(defaultDeSerializers as any),
    ...(deSerializers || {})
  };
}

// Cannot get this to work
export type CustomDeSerializer<T> =
  // infer default types
  DeSerializationMiddleware extends DeSerializationMiddleware<
    infer DefaultBinaryT,
    infer DefaultBooleanT,
    infer DefaultByteT,
    infer DefaultDecimalT,
    infer DefaultDoubleT,
    infer DefaultFloatT,
    infer DefaultInt16T,
    infer DefaultInt32T,
    infer DefaultInt64T,
    infer DefaultGuidT,
    infer DefaultSByteT,
    infer DefaultSingleT,
    infer DefaultStringT,
    infer DefaultAnyT,
    infer DefaultDateTimeT,
    infer DefaultDateTimeOffsetT,
    infer DefaultTimeT
  >
    ? DeSerializationMiddleware<
        'Edm.Binary' extends keyof T
          ? T['Edm.Binary'] extends DeSerializer<infer BinaryT> | undefined
            ? BinaryT
            : DefaultBinaryT
          : DefaultBinaryT,
        'Edm.Boolean' extends keyof T
          ? T['Edm.Boolean'] extends DeSerializer<infer BooleanT> | undefined
            ? BooleanT
            : DefaultBooleanT
          : DefaultBooleanT,
        'Edm.Byte' extends keyof T
          ? T['Edm.Byte'] extends DeSerializer<infer ByteT> | undefined
            ? ByteT
            : DefaultByteT
          : DefaultByteT,
        'Edm.Decimal' extends keyof T
          ? T['Edm.Decimal'] extends DeSerializer<infer DecimalT> | undefined
            ? DecimalT
            : DefaultDecimalT
          : DefaultDecimalT,
        'Edm.Double' extends keyof T
          ? T['Edm.Double'] extends DeSerializer<infer DoubleT> | undefined
            ? DoubleT
            : DefaultDoubleT
          : DefaultDoubleT,
        'Edm.Float' extends keyof T
          ? T['Edm.Float'] extends DeSerializer<infer FloatT> | undefined
            ? FloatT
            : DefaultFloatT
          : DefaultFloatT,
        'Edm.Int16' extends keyof T
          ? T['Edm.Int16'] extends DeSerializer<infer Int16T> | undefined
            ? Int16T
            : DefaultInt16T
          : DefaultInt16T,
        'Edm.Int32' extends keyof T
          ? T['Edm.Int32'] extends DeSerializer<infer Int32T> | undefined
            ? Int32T
            : DefaultInt32T
          : DefaultInt32T,
        'Edm.Int64' extends keyof T
          ? T['Edm.Int64'] extends DeSerializer<infer Int64T> | undefined
            ? Int64T
            : DefaultInt64T
          : DefaultInt64T,
        'Edm.Guid' extends keyof T
          ? T['Edm.Guid'] extends DeSerializer<infer GuidT> | undefined
            ? GuidT
            : DefaultGuidT
          : DefaultGuidT,
        'Edm.SByte' extends keyof T
          ? T['Edm.SByte'] extends DeSerializer<infer SByteT> | undefined
            ? SByteT
            : DefaultSByteT
          : DefaultSByteT,
        'Edm.Single' extends keyof T
          ? T['Edm.Single'] extends DeSerializer<infer SingleT> | undefined
            ? SingleT
            : DefaultSingleT
          : DefaultSingleT,
        'Edm.String' extends keyof T
          ? T['Edm.String'] extends DeSerializer<infer StringT> | undefined
            ? StringT
            : DefaultStringT
          : DefaultStringT,
        'Edm.Any' extends keyof T
          ? T['Edm.Any'] extends DeSerializer<infer AnyT> | undefined
            ? AnyT
            : DefaultAnyT
          : DefaultAnyT,
        'Edm.DateTime' extends keyof T
          ? T['Edm.DateTime'] extends DeSerializer<infer DateTimeT> | undefined
            ? DateTimeT
            : DefaultDateTimeT
          : DefaultDateTimeT,
        'Edm.DateTimeOffset' extends keyof T
          ? T['Edm.DateTimeOffset'] extends
              | DeSerializer<infer DateTimeOffsetT>
              | undefined
            ? DateTimeOffsetT
            : DefaultDateTimeOffsetT
          : DefaultDateTimeOffsetT,
        'Edm.Time' extends keyof T
          ? T['Edm.Time'] extends DeSerializer<infer TimeT> | undefined
            ? TimeT
            : DefaultTimeT
          : DefaultTimeT
      >
    : DeSerializationMiddleware;

const c = {
  'Edm.String': {
    deserialize: (val: any): number => 3,
    serialize: (val: number): any => '3'
  } as DeSerializer<number>
};

type C = CustomDeSerializer<typeof c>; // ['Edm.String'];
type D = CustomDeSerializer<typeof undefined>; // ['Edm.String'];

type Test<T extends Partial<DeSerializationMiddlewareBASE>> = T;
type TestC = Test<typeof c>;

// const d: C = {
//   'Edm.String': {
//     deserialize: (val: any): number => 3,
//     serialize: (val: number): any => '3'
//   } as DeSerializer<number>
// };

// const e: D = {
//   'Edm.String': {
//     deserialize: (val: any): string => 'tre',
//     serialize: (val: string): any => '3'
//   } as DeSerializer<string>
// };

const a = getDeSerializers({
  'Edm.String': {
    deserialize: (val: any): number => 3,
    serialize: (val: number): any => '3'
  } // as DeSerializer<number>
});

const b1 = a['Edm.String'].deserialize('1');
const b2 = a['Edm.String'].serialize(2);
const b3 = a['Edm.String'];
const b4 = a['Edm.Int32'];
