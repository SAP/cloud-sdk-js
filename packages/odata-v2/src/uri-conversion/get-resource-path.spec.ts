import { createLogger } from '@sap-cloud-sdk/util';
import { v4 as uuid } from 'uuid';
import {
  createGetResourcePathForKeys,
  createUriConverter
} from '@sap-cloud-sdk/odata-common/internal';
import { defaultDeSerializers } from '../de-serializers';
import { testEntityApi, testEntityResourcePath } from '../../test/test-util';

const uriConverter = createUriConverter(defaultDeSerializers);
const { getResourcePathForKeys } = createGetResourcePathForKeys(uriConverter);

describe('get resource path', () => {
  it('adds keys to path', () => {
    const keyPropGuid = uuid();
    const keyPropString = 'keyProp';
    const keys = {
      KeyPropertyGuid: keyPropGuid,
      KeyPropertyString: keyPropString
    };

    expect(getResourcePathForKeys(keys, testEntityApi)).toEqual(
      testEntityResourcePath(keyPropGuid, keyPropString)
    );
  });

  it('throws error if no keys set', () => {
    expect(() =>
      getResourcePathForKeys({}, testEntityApi)
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws error if not all keys are set', () => {
    const keys = { KeyPropertyGuid: uuid() };

    expect(() =>
      getResourcePathForKeys(keys, testEntityApi)
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws error if keys are nullish', () => {
    const keys = { KeyPropertyGuid: null, KeyPropertyString: undefined };

    expect(() =>
      getResourcePathForKeys(keys, testEntityApi)
    ).toThrowErrorMatchingSnapshot();
  });

  it('allows values to be empty string', () => {
    const keys = { KeyPropertyGuid: '', KeyPropertyString: '' };

    expect(() => getResourcePathForKeys(keys, testEntityApi)).not.toThrow();
  });

  it('ignores additional properties and logs a warning', () => {
    const logger = createLogger('get-resource-path');
    const warnSpy = jest.spyOn(logger, 'warn');
    const keys = {
      KeyPropertyGuid: uuid(),
      KeyPropertyString: '',
      StringProperty: 'test'
    };

    expect(() => getResourcePathForKeys(keys, testEntityApi)).not.toThrow();
    expect(warnSpy).toBeCalledWith(
      'There are too many key properties. Ignoring the following keys: StringProperty'
    );
  });

  it('URI encodes keys with special characters', () => {
    const keys = {
      KeyPropertyGuid: uuid(),
      KeyPropertyString: 'DEV?TEST06'
    };
    const expected =
      /A_TestEntity\(KeyPropertyGuid=guid'\w{8}-\w{4}-\w{4}-\w{4}-\w{12}',KeyPropertyString='DEV%3FTEST06'\)/;

    expect(getResourcePathForKeys(keys, testEntityApi)).toMatch(expected);
  });

  it('URI encodes keys containing single quote', () => {
    const keys = {
      KeyPropertyGuid: uuid(),
      KeyPropertyString: "DEV'TEST06"
    };
    const expected =
      /A_TestEntity\(KeyPropertyGuid=guid'\w{8}-\w{4}-\w{4}-\w{4}-\w{12}',KeyPropertyString='DEV''TEST06'\)/;

    expect(getResourcePathForKeys(keys, testEntityApi)).toMatch(expected);
  });

  it('URI encodes keys containing path segment separator', () => {
    const keys = {
      KeyPropertyGuid: uuid(),
      KeyPropertyString: 'DEV/TEST06'
    };
    const expected =
      /A_TestEntity\(KeyPropertyGuid=guid'\w{8}-\w{4}-\w{4}-\w{4}-\w{12}',KeyPropertyString='DEV%2FTEST06'\)/;

    expect(getResourcePathForKeys(keys, testEntityApi)).toMatch(expected);
  });
});
