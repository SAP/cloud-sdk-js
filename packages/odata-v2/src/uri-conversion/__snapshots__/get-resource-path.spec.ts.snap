// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get resource path throws error if keys are nullish 1`] = `"Cannot get resource path for entity A_TestEntity. The following keys have nullish values, but are not nullable: KeyPropertyGuid, KeyPropertyString"`;

exports[`get resource path throws error if no keys set 1`] = `"Cannot get resource path for entity A_TestEntity. The following keys are missing: KeyPropertyGuid, KeyPropertyString"`;

exports[`get resource path throws error if not all keys are set 1`] = `"Cannot get resource path for entity A_TestEntity. The following keys are missing: KeyPropertyString"`;
